Error dictionary and structured logging

ostromart opened the pull request:
Implementation of https://docs.google.com/document/d/1vdYswLQuYnrLA2fDjk6OoZx2flBABa18UjCGTn8gsg8.

istio-testing commented:
@ostromart: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/136) | `/test gencheck_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/136) | `/test gencheck_pkg`
lint_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/189) | `/test lint_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/136) | `/test gencheck_pkg`
lint_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/189) | `/test lint_pkg`
test_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/182) | `/test test_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/189) | `/test lint_pkg`
test_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/182) | `/test test_pkg`
gencheck_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/137) | `/test gencheck_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_pkg | 3b2f5b0b31200c2c952bc52a1f593b2d1ff99a57 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/182) | `/test test_pkg`
gencheck_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/137) | `/test gencheck_pkg`
lint_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/190) | `/test lint_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/137) | `/test gencheck_pkg`
lint_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/190) | `/test lint_pkg`
test_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/183) | `/test test_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
```
Are there logging ingestion tools that read logs that are partially plaintext and partially json?

ostromart replied:
Yes, but it's up to us to decide exactly how the log outputs. It could be all JSON, text + JSON or all text. That can easily be made more configurable.

howardjohn replied:
how come in the previous test its json with spaces and in this one there are no spaces?

howardjohn commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	var ss []string
+	if ie != nil {
+		ss = append(ss, fmt.Sprintf(`"moreInfo":"%s"`, ie.MoreInfo))
+		ss = append(ss, fmt.Sprintf(`"impact":"%s"`, ie.Impact))
```
hand crafting json with arbitrary inputs means the output may easily become invalid json

ostromart replied:
i think we can do better here to avoid that. i'd rather not include a json library to format the json though. 
how about, number types have no quotes, everything else has escaped quotes? then objects will not cause invalid json, only messy json :-)

howardjohn commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/log"
```
Where does this come from?

ostromart replied:
from goland being helpful :-)

ostromart replied:
removed.

howardjohn commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	for _, zfi := range fields {
+		zf, ok := zfi.(zapcore.Field)
+		if !ok {
+			log.Errorf("bad interface type: expect zapcore.Field, got %T for fields %v", zf, fields)
```
This should definitely not be here (it is the Prometheus log library)

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `{"moreInfo":"MoreInfo","impact":"Impact","action":"Action","likelyCauses":"LikelyCause","foo":"bar","message":"Hello"}`)
```
Why is message last? It seems the most important

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
 }
 
+func TestScopeAddLabel(t *testing.T) {
+	const name = "TestScope"
```
I am very confused how I am supposed to use the library now. Seems like we have 3 distinct ways to add metadata (scope, label, context) with no clear distinction as to when I should use one or the other

ostromart replied:
See comment about label/context - we can remove one of these if it's going to be too confusing when to use one or the other. 
I think the difference between scope and label/context should be clear though - labels/context are bound to a scope. I agree more examples/docs are needed.

ostromart replied:
removed context for now. we can discuss what to do with it in a chat with doug.

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+
+	expect := `Hello {"foo":"bar","baz":1,"qux":0.123}`
+	mustRegexMatchString(t, lines[0], expect)
+	mustRegexMatchString(t, lines[1], expect)
```
why is there 2 lines?
Why is it a regex match instead of exact match - from the timestamp?

douglas-reid commented on log/scope.go:
```diff
@@ -315,3 +343,149 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+		return nil, 0
+	}
+	// Skip IstioErrorStruct, pass remaining fields on as before.
+	return ie, 1
```
clarifying: to use the structured error, it *must* be the first argument in the `Debuga` (or similar) call?  Is the intent to make the API the same while adding the functionality, then?

Is adding a bunch of `DebugStructured()`, etc.,  that enforce the ordering directly problematic because of all the levels ?

ostromart replied:
Right, exactly. An alternative would be to name the calls something different (InfoStructf, DebugStructf...). 
I'm ok with either approach. The second approach would be a brand new thing rather than trying for backward compatibility which could actually be a good thing.

douglas-reid commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `{"moreInfo":"MoreInfo","impact":"Impact","action":"Action","likelyCauses":"LikelyCause","foo":"bar","message":"Hello"}`)
```
i'm a little nervous that the output changes format based on the inclusion of the `ie` in the call to `Debuga()`. That seems like it might surprise.

Somewhat related, I don't know what the right thing here is, but I worry that we'll have alternating log lines like:

```text
Hello {"foo": "bar", "baz": "goo"}
{"moreInfo"... "message":"Hello"}
Goodbye {"foo"...}
{"moreInfo"..."message":"Goodbye"}
```

This seems potentially like a mess to parse. Am I reading this incorrectly?

ostromart replied:
Good point, how about:
```
Hello {"foo": "bar", "baz": "goo"}
Hello { "message":"Hello","moreInfo"...}
```

douglas-reid replied:
i'm fairly torn on this. what does zap default to?  does the console encoder thingy use half text, half JSON?  if so, then we should probably match that, and then only switch to full JSON if that's requested in the config (which I believe is your proposal).

As an aside, I almost feel like cli-style (`Hello foo=bar baz=goo`) or full JSON (`{"message":"Hello", "foo":"bar", ...}`) are better than mixing JSON and free text, but that seems like an issue for zap, not Istio.

ostromart commented on errdict/errdict.go:
```diff
@@ -0,0 +1,19 @@
+
+// IstioErrorStruct represents structured error information, for optional use in scope.X or log.X calls.
+// See https://docs.google.com/document/d/1vdYswLQuYnrLA2fDjk6OoZx2flBABa18UjCGTn8gsg8/ for additional information.
+type IstioErrorStruct struct {
```
All of those are better than IstioErrorStruct - I vote for structured.Error. btw. the reason for the Struct at the end was that I was planning to wrap this in an interface, but then I decided the extra complexity is not worth it.

ostromart commented on errdict/errdict.go:
```diff
@@ -0,0 +1,19 @@
+// See https://docs.google.com/document/d/1vdYswLQuYnrLA2fDjk6OoZx2flBABa18UjCGTn8gsg8/ for additional information.
+type IstioErrorStruct struct {
+	// MoreInfo is additional information about the error e.g. a link to context describing the context for the error.
+	MoreInfo string
```
It's in the original message text itself i.e. what we would print today. We clearly need more docs and examples here but I wanted to get basic consensus before generating a ton of docs for this.

ostromart commented on errdict/errdict.go:
```diff
@@ -0,0 +1,19 @@
+	// Action is the next step the user should take e.g. "Open an issue or bug report."
+	Action string
+	// LikelyCause is the likely cause for the error e.g. "Software bug."
+	LikelyCause string
```
Not yet, but the idea would be to have some consts defined in the dictionary package. I also considered having this as a slice but it gets too unwieldly.

ostromart commented on errdict/errdict.go:
```diff
@@ -0,0 +1,19 @@
+	LikelyCause string
+
+	// Metrics is a list of associated metrics that will be passed to a handler from the call site.
+	Metrics []monitoring.Metric
```
There's an example/description in the doc and I will copy it here if that looks good?

ostromart commented on log/scope.go:
```diff
@@ -315,3 +343,149 @@ func (s *Scope) SetLogCallers(logCallers bool) {
 }
+
+// RegisterHandler registers a handler for s that is called for all Fatal*, Error*, Warn*, Info* and Debug* function calls.
+func (s *Scope) RegisterHandler(callback ScopeHandlerCallbackFunc) {
```
It's meant to be used to pass the same structured info (labels, context) primarily also to a metrics handler in the future. There could also be a couple of different flavor of text loggers (e.g. performance vs. nice output).
 I think there's very low likelihood of anyone except us adding any handlers but we could make this private interface for now.

ostromart replied:
removed the scope version. let's stick with the global handlers to limit the scope for bad things to happen.

ostromart commented on log/scope.go:
```diff
@@ -315,3 +343,149 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+}
+
+// WithContext returns a copy of s with the given key-value appended to the context.
+func (s *Scope) WithContext(key string, value interface{}) *Scope {
```
See the unit tests. Add* means it persists for future scope calls, With* is for this call alone.

ostromart replied:
removed this in favor of copying.

ostromart commented on log/scope.go:
```diff
@@ -315,3 +343,149 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+}
+
+// AddLabel adds a key-value pair to the labels in s.
+func (s *Scope) AddLabel(key string, value interface{}) *Scope {
```
context just adds some OT compliant formatting and it's meant for things like transaction ID. not sure that we need both though, they are very similar in output - thoughts @douglas-reid ?

ostromart commented on log/scope_test.go:
```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123).AddLabel("qux", 0.123).Debuga("Hello")
+		s.Debuga("Hello")
```
WithLabel would not apply to either call, only if we had s.WithLabel().Info(). 
We need something that controls "for this debug statement only" vs. "add to the scope for every subsequent debug statement" - any suggestions how that should look? I don't think assignment can be used to trigger afaik in Go.

howardjohn replied:
I would suggest always copying. If you want to persist the changes you do `scope = scope.WithLabels()`. This seems more explicit/idiomatic, and then we only need a single function?

ostromart replied:
ok, agreed. i was worried about having too many copies but this is way simpler and in reality i think the copies are not a problem.

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	fields []zapcore.Field) {
+	var ss []string
+	if ie != nil {
+		ss = append(ss, fmt.Sprintf(`"moreInfo":"%s"`, ie.MoreInfo))
```
this will be configurable. the plan was to add that in a follow-on PR but i will include it in here.

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+		ss = append(ss, fmt.Sprintf(`"likelyCauses":"%s"`, ie.LikelyCause))
+	}
+	if len(scope.structuredKey) > 0 {
+		scope.mu.RLock()
```
ack, per offline discussion this was to avoid copies but if copies are ok we can get rid of this. i will update with the locks removed and a few more copies.

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	}
+	if len(ss) > 0 {
+		ss = append(ss, fmt.Sprintf(`"message":"%s"`, msg))
+		msg = "{" + strings.Join(ss, ",") + "}"
```
i think it does warrant some explanation at the file comment level. zap structured logging is not much more than what is here and it seems pointless to conform this to zap structures just to leverage that tiny bit of code. the structured logging is actually not more performant - the lean and mean log implementation is in the "emit" call which we still retain.

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	return zfs
+}
+
+const callerSkipOffset = 4
```
there wasn't in the original code i got it from but i will add one.

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	}
+}
+
+// backwards compatibility. TODO(mostrowski): remove this
```
it's to preserve the semantics in the default.go file. i'll get rid of it shortly.

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/137) | `/test gencheck_pkg`
test_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/183) | `/test test_pkg`
lint_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/194) | `/test lint_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
test_pkg | dcf3f02eab7cc54f5e6313f9fb3d9a84f3b3255d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/183) | `/test test_pkg`
lint_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/194) | `/test lint_pkg`
gencheck_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/141) | `/test gencheck_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/194) | `/test lint_pkg`
gencheck_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/141) | `/test gencheck_pkg`
test_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/187) | `/test test_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ostromart commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,129 @@
+	}
+	if len(ss) > 0 {
+		ss = append(ss, fmt.Sprintf(`"message":"%s"`, msg))
+		msg = "{" + strings.Join(ss, ",") + "}"
```
i took john's suggestion to just use zap. since zap is all about performance i trust all their type reflection is blazing fast :-)

ostromart commented:
@howardjohn @douglas-reid ptal a look again, i've incorporated most of your suggestions:
 - removed context (for now, we can put it back)
 - all scope functions return a copy so i could remove the locking and the semantics are simpler now
 - refactored the structured output to use zap fields (and hence their JSON encoder)
 - renamed errdict.IstioErrorStruct to structured.Error
 - updated the field ordering to put the message first
 - message is always output outside the JSON block for consistency. 
 - message is the first item in the JSON block
if you guys are ok with the results i'll start work on documentation.

howardjohn commented on go.mod:
```diff
@@ -25,6 +25,7 @@ require (
 	github.com/onsi/gomega v1.5.0
 	github.com/pkg/errors v0.8.1 // indirect
 	github.com/prometheus/client_golang v0.9.4
+	github.com/prometheus/common v0.4.1
```
this is suspicious - hopefully the test fails

ostromart replied:
done.

howardjohn commented on log/scope.go:
```diff
@@ -39,10 +38,37 @@ type Scope struct {
+
+	// context data - key slice to preserve ordering
+	contextKey []string
+	context    map[string]interface{}
```
I thought we got rid of context?

ostromart replied:
forgot to remove this bit, done.

howardjohn commented on log/scope.go:
```diff
@@ -39,10 +38,37 @@ type Scope struct {
+
+// ScopeHandlerCallbackFunc is a callback type for the handler called from Fatal*, Error*, Warn*, Info* and Debug*
+// function calls.
+type ScopeHandlerCallbackFunc func(
```
nit: private type?

ostromart replied:
good idea, done.

howardjohn commented on log/scope.go:
```diff
@@ -39,10 +38,37 @@ type Scope struct {
+
+// RegisterDefaultHandler registers a scope handler that is called by default from all scopes. It is appended to the
+// current list of scope handlers.
+func RegisterDefaultHandler(callback ScopeHandlerCallbackFunc) {
```
nit: private function?

howardjohn replied:
I care a decent amount about making everything in log that can be private, private. To avoid confusion

ostromart replied:
done.

howardjohn commented on log/scope.go:
```diff
@@ -130,27 +165,33 @@ func (s *Scope) FatalEnabled() bool {
 
 // Errorf uses fmt.Sprintf to construct and log a message at error level.
-func (s *Scope) Errorf(template string, args ...interface{}) {
+// func (s *Scope) Errorf(template string, args ...interface{}) {
```
nit: stray comments

ostromart replied:
done.

howardjohn commented on log/scope.go:
```diff
@@ -130,27 +165,33 @@ func (s *Scope) FatalEnabled() bool {
 // Errorf uses fmt.Sprintf to construct and log a message at error level.
-func (s *Scope) Errorf(template string, args ...interface{}) {
+// func (s *Scope) Errorf(template string, args ...interface{}) {
+func (s *Scope) Errorf(args ...interface{}) {
```
This *might* be a backwards incompatible API change. Or maybe not. But why do we need it? Is it valid to pass `Errorf(<not string>, something)`? What happens in this case?

ostromart replied:
we need it to retain the familiar interface to scope.Debug* et al. while allowing structured.Error to be optionally passed. the alternative (per thread with doug) is to call the functions taking structured.Error something else. but what? i feel it could get messy to have a proliferation of these functions with slightly different names.

howardjohn commented on log/scope.go:
```diff
@@ -315,3 +338,96 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+
+// getErrorStruct returns (*Error, 1) if it is the first argument in the list is an Error ptr,
+// or (nil,0) otherwise. The second return value is the offset to the first non-Error field.
+func getErrorStruct(fields ...interface{}) (*structured.Error, int) {
```
nit: why even return int here? seems like callers can just check if the response is == nil. Its internal though so NBD

ostromart replied:
just to avoid checking every time at the caller site. makes the code more compact for the 50 wrapper functions.

howardjohn replied:
ack, thanks

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,89 @@ func TestBasicScopes(t *testing.T) {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello	{"message": "Hello", "moreInfo": "MoreInfo", "impact": "Impact", "action": "Action", "likelyCauses": "LikelyCause", "foo": "bar"}`)
```
Where does the big space between Hello and `{` come from? Or is a tab and it displays weird on the diff?

ostromart replied:
these are both due to switching to the zap json encoder. it's their fault now!

howardjohn commented on log/scope_test.go:
```diff
@@ -218,6 +220,89 @@ func TestBasicScopes(t *testing.T) {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello	{"message": "Hello", "moreInfo": "MoreInfo", "impact": "Impact", "action": "Action", "likelyCauses": "LikelyCause", "foo": "bar"}`)
```
Do we need the message in two places? Seems a bit redundant. I guess for log processors they just look at the json part?

ostromart replied:
per doug's point, we need to be consistent. but we can debate exactly what format we want to emit and it's easy to change it. if we vote for either cli or json and not a mix i'm ok with that.

ostromart replied:
fyi. zap does a mix of tabbed and json (for the fields) but i agree with doug that this is inconsistent.

howardjohn commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,123 @@
+	msg string,
+	fields []zapcore.Field) {
+	if ie != nil {
+		fields = append(fields, zap.String("message", msg))
```
We currently have a `--log_as_json` option. What happens when this is enabled and we have structured logging? hopefully not double nested json?

howardjohn replied:
As a further point, does it make sense to output json only if that flag is enabled, and if its disabled something like `Hello moreInfo=some more info impact=some impact action=some action likelyCauses=some cause`? Similar to https://github.com/sirupsen/logrus format I guess

ostromart replied:
yes, will make this change in a bit.

ostromart replied:
done.

howardjohn commented on log/zapcore_handler.go:
```diff
@@ -0,0 +1,123 @@
+	if ie != nil {
+		fields = append(fields, zap.String("message", msg))
+		fields = append(fields, zap.String("moreInfo", ie.MoreInfo))
+		fields = append(fields, zap.String("impact", ie.Impact))
```
should we append only if they are not empty?

ostromart replied:
done.

howardjohn commented on log/scope.go:
```diff
@@ -315,3 +338,96 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+func (s *Scope) WithLabels(kvlist ...interface{}) *Scope {
+	out := s.Copy()
+	if len(kvlist)%2 != 0 {
+		out.labels["WithLabels error"] = fmt.Sprintf("even number of parameters required, got %d", len(kvlist))
```
Just curious what `zap` does here? I can't think of a better way to do things, just curious if someone from zap has

ostromart replied:
they use fprintf :-D

istio-testing commented:
@ostromart: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/gencheck_pkg/141) | `/test gencheck_pkg`
test_pkg | e98bd0ceb17d760dc6ef1bcb4cdffa74354f436c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/test_pkg/187) | `/test test_pkg`
lint_pkg | 0729cd4bf4c3d27dfc98929d6abdf3cc8a40adf9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/195) | `/test lint_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ostromart: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_pkg | b05f95b3fa582794d64d013120b55d8d4b4775c2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_pkg/192/lint_pkg/196) | `/test lint_pkg`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented on structured/structured.go:
```diff
@@ -0,0 +1,35 @@
+	LikelyCause string
+
+	// Metrics is a list of associated metrics that will be passed to a handler from the call site.
+	Metrics []monitoring.Metric
```
since the `Metrics` part doesn't appear in any tests, and isn't used yet, should we remove it from the initial PR in the goal of starting simple, or do we feel like it will be used in short order?

ostromart replied:
unfortunately the deadline is today so there wouldn't be time to do it separately. yes, i will put up some samples pretty quickly afterwards (cleaning logs and adding metrics shouldn't be subject to code freeze).

howardjohn commented on log/scope.go:
```diff
@@ -100,27 +144,30 @@ func Scopes() map[string]*Scope {
 }
 
 // Fatala uses fmt.Sprint to construct and log a message at fatal level.
 func (s *Scope) Fatala(args ...interface{}) {
```
I am not sure how this differs from `Fatal` in terms of output. I think with your interface changes it may actually be identical? Can we verify and if so as a followup remove the `*a` ones?

ostromart replied:
Fatala doesn't take structured parameters to pass to zap. If no one uses those then we can remove (actually I'd replace Fatal with the Fatala implementation and remove Fatala).

howardjohn commented on log/scope.go:
```diff
@@ -315,3 +347,103 @@ func (s *Scope) SetLogCallers(logCallers bool) {
 }
+
+// Copy makes a copy of s and returns a pointer to it.
+func (s *Scope) Copy() *Scope {
```
Is there a use case for this or should it be private?

ostromart replied:
yes, thanks for pointing out, it should be now.

howardjohn commented on log/scope.go:
```diff
@@ -315,3 +347,103 @@ func (s *Scope) SetLogCallers(logCallers bool) {
+// e.g. myScope = myScope.WithLabels("foo", "bar", "baz", 123, "qux", 0.123)
+//      myScope = myScope.WithoutLabels("baz", "qux")
+// Result: myScope retains only the "foo" label.
+func (s *Scope) WithoutLabels(keys ...string) *Scope {
```
nit: do we need this? I would expect usage to be

```
s := some-scope
if ... {
  s = s.WithLabels(foo)
  s.Log(something with foo)
}
s.Log(something without foo)
```

howardjohn replied:
Basically I am biased in favor of minimal API until proven a need otherwise. We can always add, hard to remove and use a big API though. But no strong opinion here

ostromart replied:
that's a good question. i guess i took it for granted that someone would want to do this but when pressed i'm having a hard time coming up with a convincing example. i will remove for now, we can indeed easily add it back.

istio-testing commented:
In response to a cherrypick label: new pull request created: #197
